File: fieldmap-app/src/models/FieldmapFile.ts
import type { Pin } from './Pin';
import type { Track } from './Track';
import type { Layer } from './Layer';
import type { Project } from './Project';

/** Format version — bump when the on-disk shape changes incompatibly. */
export const FIELDMAP_FILE_VERSION = 1;

/**
 * The .fieldmap export format. Single JSON file containing everything
 * a project needs to be opened on another device.
 *
 * Mirrors FieldmapFile.swift.
 */
export interface FieldmapFile {
  version: number;
  exportedAt: string;
  exportedBy: string; // app version, e.g. "Fieldmap 0.1.0"
  project: Project;
  pins: Pin[];
  tracks: Track[];
  layers: Layer[];
}

File: _archive/Fieldmap-Xcode-Project/Fieldmap/Models/FieldmapFile.swift
import Foundation

/// The .fieldmap file format — a JSON container for a whole project
struct FieldmapFile: Codable {
    let version: String
    let exportedAt: Date
    let project: Project
    let layers: [Layer]
    let pins: [Pin]
    let tracks: [Track]
    
    init(project: Project, layers: [Layer], pins: [Pin], tracks: [Track]) {
        self.version = "1.0"
        self.exportedAt = Date()
        self.project = project
        self.layers = layers
        self.pins = pins
        self.tracks = tracks
    }
    
    /// Encode to JSON data
    func toJSON() throws -> Data {
        let encoder = JSONEncoder()
        encoder.dateEncodingStrategy = .iso8601
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        return try encoder.encode(self)
    }
    
    /// Decode from JSON data
    static func fromJSON(_ data: Data) throws -> FieldmapFile {
        let decoder = JSONDecoder()
        decoder.dateDecodingStrategy = .iso8601
        return try decoder.decode(FieldmapFile.self, from: data)
    }
}
